### PR TITLE
00387: CVE-2020-10735: Prevent DoS by very large int()

### PR DESCRIPTION
**Pull request by vstinner**
gh-95778: CVE-2020-10735: Prevent DoS by very large int() (GH-96504)

Converting between `int` and `str` in bases other than 2 (binary), 4, 8 (octal), 16 (hexadecimal), or 32 such as base 10 (decimal) now raises a `ValueError` if the number of digits in string form is above a limit to avoid potential denial of service attacks due to the algorithmic complexity. This is a mitigation for CVE-2020-10735 (https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-10735).

This new limit can be configured or disabled by environment variable, command line flag, or :mod:`sys` APIs. See the `Integer String Conversion Length Limitation` documentation.  The default limit is 4300 digits in string form.

Patch by Gregory P. Smith [Google] and Christian Heimes [Red Hat] with feedback from Victor Stinner, Thomas Wouters, Steve Dower, Ned Deily, and Mark Dickinson.

Notes on the backport to Python 3.6:

* Use "Python 3.6.16" version in the documentation, whereas this version will never be released
* Only add _Py_global_config_int_max_str_digits global variable: Python 3.6 doesn't have PyConfig API (PEP 597) nor _PyRuntime.
* sys.flags.int_max_str_digits cannot be -1 on Python 3.6: it is set to the default limit. Adapt test_int_max_str_digits() for that.
* Declare _PY_LONG_DEFAULT_MAX_STR_DIGITS and _PY_LONG_MAX_STR_DIGITS_THRESHOLD macros in longobject.h but only if the Py_BUILD_CORE macro is defined.
* Declare _Py_global_config_int_max_str_digits in pydebug.h.

### CODE REVIEW

**Comment by vstinner** [comment]
@corona10 @mdickinson: Here is my backport of the CVE-2020-10735 to Python 3.6.

**Comment by mdickinson** [comment]
@vstinner Thank you for the link! This is very helpful.

File: Doc/library/functions.rst [hunk 1]
```diff
@@ -747,6 +747,14 @@ are always available.  They are listed here in alphabetical order.
 
    .. versionchanged:: 3.6
       Grouping digits with underscores as in code literals is allowed.
+ 
+   .. versionchanged:: 3.6.16
```
Comment (hroncok) [inline Doc/library/functions.rst]: 3.6.16 will never exist. For Fedora purposes, we can use version-release, like this:

```suggestion
   .. versionchanged:: 3.6.15-12
```

Comment (vstinner) [reply Doc/library/functions.rst]: Oh right, I replaced version with ``3.6.15-13``.

**Comment by vstinner** [comment]
Update:

* Replace (non existent) version ``3.6.16`` with (Fedora package version) ``3.6.15-13``
* Backport error message fix (commit e841ffc915e82e5ea6e3b473205417d63494808d)
* Backport command line parsing fix (commit 41351662bcd21672d8ccfa62fe44d72027e6bcf8)

**Comment by vstinner** [comment]
PR rebased on top of 3.6.15-12 fix.

**Comment by stratakis** [comment]
Everything looks good, feel free to force-push.

**Comment by vstinner** [comment]
I pushed the commit (``git push origin HEAD``) and created ``fedora-3.6.15-13`` tag.

**Pull request merged**